### PR TITLE
Fixes evanphx/json-patch#128 - custom marshalling for doc nodes

### PR DESCRIPTION
**Pull request by mrpotes**
Rather than using a `map[string]*lazyNode` for the doc (object) nodes, this is changed to use a rich type with both a map and an array of key-value pairs, and custom `MarshalJSON` and `UnmarshalJSON` functions.

### CODE REVIEW

File: v5/go.sum
Comment (mrpotes) [inline v5/go.sum:6]: I think this can be reverted.

File: v5/patch.go
Comment (mrpotes) [inline v5/patch.go:59]: It occurs to me that I don't actually need the `docItem` type, as all that is needed is the order of the keys - the values are already present in the `obj` map. I'm happy to update this.

File: v5/patch.go
Comment (evanphx) [inline v5/patch.go:132]:
```suggestion
	var buf bytes.Buffer
```

File: v5/patch.go
Comment (evanphx) [inline v5/patch.go:199]: This looks fine but it makes me want to submit a patch to `encoding/json` to support some sort of `[]KeyValue` type for unmarshaling a JSON object so we don't have to do it....

Maybe I'll write that up.

**Review by evanphx — APPROVED** [review]
One super minor thing, otherwise looks great and I'll merge it after you fix the one tiny thing.

File: v5/patch.go
Comment (mrpotes) [reply v5/patch.go]: https://github.com/golang/go/issues/27179 discusses how they might sort this out in `encoding/json`

File: v5/patch.go
Comment (mrpotes) [reply v5/patch.go:199]: golang/go#27179 discusses how they might sort this out in `encoding/json`

Comment (evanphx) [reply v5/patch.go:199]: Ah! Thanks you for that! A wonderful resource for writing up a PR.

**Comment by mrpotes** [comment]
Thanks for the suggestion, committed it - do you want the commits squashed, or will you do that as part of merging?

**Comment by evanphx** [comment]
I can squash them here, no prob. Once CI passes, I'll merge.

**Comment by evanphx** [comment]
Travis is busted, merging without CI and retesting.

**Pull request merged**
